### PR TITLE
Add auto-bonding feature

### PR DESCRIPTION
**Pull request by almusil**
The labeler will now look through the lldp system name of the interface
switches. If there is match the labeler will try to create
active backup bond between those interfaces.

Signed-off-by: Ales Musil <amusil@redhat.com>

### CODE REVIEW

**Comment by dankenigsberg** [comment]
Please make sure that the new auto-bonder is separately configured: a user may want auto label, but not auto-bonding and vice versa.

Also, we should create LACP (mode 4) bonds, and only on nics that belong to the same PortGroup on the switch. This is the most delicate and error-prone task that the auto bonder needs to automate.

When you make an action, please emit a very clear log, e.g. found eth0, eth1, eth2 reporting lldp portgroup 7. Creating a bond on top of them.

When one of the nics already have a network on it, you should move the network to the new bond. otherwise, bonding action would fail.

if different network exist on different candidate slaves, we'd better not even attempt bonding. We should also think twice whether we want to auto-bond nics that carry the management network.

(I am writing down comments after talking to @mburman5 )

File: src/labeler/labeler.py [hunk 1]
```diff
@@ -43,88 +43,143 @@ def _get_host_interfaces_service(host_id):
...
     try:
         nic_service = nics_service.nic_service(nic.id)
         lldp_list = nic_service.link_layer_discovery_protocol_elements_service().list()
-        if vlan_only:
-            lldp_list = utils.filter_vlan_tag(lldp_list)
+        logging.info('Getting lldp information for nic %s', nic.name)
         return lldp_list
     except api.sdk.Error as ex:
-        print(ex.message)
+        logging.info(ex.message)
```
Comment (maiqueb) [inline src/labeler/labeler.py]: shouldn't this log level be error? warn at least?

File: README.adoc [hunk 2]
```diff
@@ -61,6 +61,11 @@ _default:_ https://ovirt-engine/ovirt-engine/api
 ca_file:: Path to the custom certificate file. Leave empty if you don't want to use custom certificate. +
 _default:_
 
+auto_bonding:: Boolean value (true/false). If it set to true the labeler will do auto bonding. +
```
[the remark this answers is not in the document]
Comment (almusil) [reply README.adoc]: Done

File: README.adoc [hunk 3]
```diff
@@ -61,6 +61,11 @@ _default:_ https://ovirt-engine/ovirt-engine/api
...
+_default:_ true
+
+auto_labeling:: Boolean value (true/false). If it set to true the labeler will do auto labeling. +
```
[the remark this answers is not in the document]
Comment (almusil) [reply README.adoc]: Done

File: src/labeler/labeler_utils.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+
+def check_bond_slaves_attachments(nic_list, network_dict):
+    checked_nic_list = []
+    has_non_vlan_network = False
+    for nic in nic_list:
+        networks = network_dict.get(nic, None)
+        if networks is not None and not _contains_managment_network(networks):
+            contains_non_vlan_network = _contains_non_vlan_network(networks)
+            if contains_non_vlan_network and not has_non_vlan_network:
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler_utils.py]: I did a small refactor to make it more readable hopefully.

File: src/labeler/labeler_utils.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+#
+# Refer to the README and COPYING files for full details of the license
+import logging
+import re
+
+from ovirtsdk4.types import Bonding, HostNic, Option
+
+LINK_AGGREGATION_TYPE = 127
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler_utils.py]: I will do probably in some follow up patch

File: src/labeler/labeler_utils.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        if networks is not None:
+            network_list.extend(networks)
+    return network_list
+
+
+def create_attachment_definition(nic_list, next_bond_num, attachment_dict):
+    attachments_list = []
+    bond_name = BOND_PREFIX + str(next_bond_num)
+    for nic in nic_list:
+        attachments = attachment_dict.get(nic, None)
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler_utils.py]: Done

File: src/labeler/labeler_utils.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        logging.info('Creating bond %s with slaves: %s', bond_name, ', '.join([nic.name for nic in nic_list]))
+        return HostNic(name=bond_name, bonding=bonding)
+    else:
+        return None
+
+
+def create_network_list(nic_list, network_dict):
+    network_list = []
+    for nic in nic_list:
+        networks = network_dict.get(nic, None)
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler_utils.py]: Done

File: src/labeler/labeler_utils.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+
+def update_bond_dict(bond_dict, tlvs, nic):
+    link_aggregation_properties = filter_link_aggregation(tlvs)
+    if link_aggregation_properties is not None:
+        aggregation_port_id = \
+            (prop.value for prop in link_aggregation_properties if prop.name == PROPERTY_LINK_AGGREGATION_ID).next()
+        if aggregation_port_id is None:
+            return
+        slave_list = bond_dict.get(aggregation_port_id, None)
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler_utils.py]: I am not sure how to do the check then. I need to know if there is existing list just append to it. If there is none create new one.

File: src/labeler/labeler_utils.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    else:
+        bond_names.sort()
+        return int(bond_names[0].strip('bond')) + 1
+
+
+def update_bond_dict(bond_dict, tlvs, nic):
+    link_aggregation_properties = filter_link_aggregation(tlvs)
+    if link_aggregation_properties is not None:
+        aggregation_port_id = \
+            (prop.value for prop in link_aggregation_properties if prop.name == PROPERTY_LINK_AGGREGATION_ID).next()
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler_utils.py]: Yes. Hopefully I have fixed the indentation a little bit.

File: src/labeler/labeler_utils.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+            slave_list.extend([slave.id for slave in nic.bonding.slaves])
+    return [nic for nic in nic_list if nic.id not in slave_list]
+
+
+def filter_link_aggregation(tlvs):
+    link_aggregation = None
+    for tlv in tlvs:
+        if tlv.type == LINK_AGGREGATION_TYPE and tlv.oui == LINK_AGGREGATION_OUI \
+                and tlv.subtype == LINK_AGGREGATION_SUBTYPE:
+            link_aggregation = tlv.properties
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler_utils.py]: It actually returns list of properties.

File: src/labeler/labeler_utils.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+def _is_currently_aggregated(properties):
+    prop_val = [prop.value for prop in properties if (prop.name == PROPERTY_LINK_AGGREGATION_AGGREGATED
+                                                      and prop.value == PROPERTY_TRUE)]
+    return len(prop_val) > 0
+
+
+def find_next_bond_num(nic_names):
+    bond_names = [nic_name for nic_name in nic_names if re.search(r'bond.*', nic_name)]
+    if len(bond_names) == 0:
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler_utils.py]: Done

File: src/labeler/labeler.py [hunk 12]
```diff
@@ -43,88 +43,143 @@ def _get_host_interfaces_service(host_id):
...
+        elif not without_bonds:
+            lldps.update({nic: _get_lldp_for_bond_slaves(nics_service, nic)})
     return lldps
 
 
+def _get_networks_from_attachments(nic_list, attachment_dict):
+    networks_service = _get_engine_service().networks_service()
+    network_dict = {}
+    for nic in nic_list:
+        network_attachments = attachment_dict.get(nic, None)
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler.py]: Done

File: src/labeler/config.py.in [hunk 13]
```diff
@@ -58,6 +61,14 @@ def get_ca_file():
     return _config.get(_LABELER_SECTION, _CA_FILE_CONFIG)
 
 
+def get_auto_bonding():
+    return False if _config.get(_LABELER_SECTION, _AUTO_BONDING_CONFIG) != 'true' else True
+
+
+def get_auto_labeling():
+    return False if _config.get(_LABELER_SECTION, _AUTO_LABELING_CONFIG) != 'true' else True
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/config.py.in]: Done

File: src/labeler/config.py.in [hunk 14]
```diff
@@ -58,6 +61,14 @@ def get_ca_file():
     return _config.get(_LABELER_SECTION, _CA_FILE_CONFIG)
 
 
+def get_auto_bonding():
+    return False if _config.get(_LABELER_SECTION, _AUTO_BONDING_CONFIG) != 'true' else True
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/config.py.in]: Done

File: src/labeler/labeler_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (maiqueb) [reply src/labeler/labeler_utils.py]: fair enough.

[the remark this answers is not in the document]
Comment (maiqueb) [reply src/labeler/labeler_utils.py]: fair enoguh

File: src/labeler/labeler_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (maiqueb) [reply src/labeler/labeler_utils.py]: ah. did not see that coming.

File: src/labeler/labeler_utils.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (maiqueb) [reply src/labeler/labeler_utils.py]: way better. i still don't like it, but, it's way better. no arguing about that.

File: src/labeler/labeler_utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (maiqueb) [reply src/labeler/labeler_utils.py]: got it. i just loathe None / null and the likes of it. 

you could do the same check with an empty list, but, I agree doing that would be silly - would not be consistent with the rest of the code.

File: src/labeler/labeler_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (maiqueb) [reply src/labeler/labeler_utils.py]: i find it almost the same - the new variable name helps though.

File: src/labeler/labeler.py [hunk 15]
```diff
@@ -43,88 +43,142 @@ def _get_host_interfaces_service(host_id):
...
+        network_dict.update({nic: [networks_service.network_service(attachment.network.id).get() for attachment
+                            in network_attachments]})
+    return network_dict
+
+
+def _create_new_bonds(bond_dict, attachment_dict, host_id, next_bond_num):
+    bonds_to_update = []
+    attachments_to_update = []
+    for aggregation_id, nic_list in bond_dict.items():
+        network_dict = _get_networks_from_attachments(nic_list, attachment_dict)
```
Comment (dominikholler) [inline src/labeler/labeler.py]: I wonder if we should be more conservative here, e.g. bond only NICs with no network attached?

File: Makefile [hunk 16]
```diff
@@ -15,7 +15,7 @@
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 #
 # Refer to the README and COPYING files for full details of the license
-VERSION=1.0.0
+VERSION=1.0.1
```
Comment (dominikholler) [inline Makefile]: Maybe `1.0.1-SNAPSHOT` if this is no released version.

File: README.adoc [hunk 17]
```diff
@@ -61,6 +61,11 @@ _default:_ https://ovirt-engine/ovirt-engine/api
 ca_file:: Path to the custom certificate file. Leave empty if you don't want to use custom certificate. +
 _default:_
 
+auto_bonding:: Boolean value (true/false). If set to true the labeler will do auto bonding. +
```
[the remark this answers is not in the document]
Comment (almusil) [reply README.adoc]: Done

File: README.adoc [hunk 18]
```diff
@@ -61,6 +61,11 @@ _default:_ https://ovirt-engine/ovirt-engine/api
...
+_default:_ true
+
+auto_labeling:: Boolean value (true/false). If set to true the labeler will do auto labeling. +
+_default:_ true
 # ovirt-lldp-labeler.timer
```
[the remark this answers is not in the document]
Comment (almusil) [reply README.adoc]: Done

File: src/labeler/labeler_utils.py [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+def filter_out_bond_slaves(nic_list):
+    slave_list = []
+    for nic in nic_list:
+        if nic.bonding is not None:
+            slave_list.extend([slave.id for slave in nic.bonding.slaves])
+    return [nic for nic in nic_list if nic.id not in slave_list]
+
+
+def filter_link_aggregation(tlvs):
+    for tlv in tlvs:
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler_utils.py]: Good point. The case still seems pretty rare to me. But I will address that in follow up patch

File: src/labeler/labeler.py [hunk 20]
```diff
@@ -43,88 +43,142 @@ def _get_host_interfaces_service(host_id):
...
-    nic_list_without_vlan = _filter_out_vlan_interfaces(nics_service.list())
-    nic_list = _filter_out_bond_slaves(nic_list_without_vlan)
+    nic_list_without_vlan = utils.filter_out_vlan_interfaces(nics_service.list())
+    nic_list = utils.filter_out_bond_slaves(nic_list_without_vlan)
     for nic in nic_list:
         if nic.bonding is None:
-            lldps.update({nic.id: _get_lldp_for_nic(nics_service, nic)})
-        else:
-            lldps.update({nic.id: _get_lldp_for_bond_slaves(nics_service, nic)})
+            lldps.update({nic: _get_lldp_for_nic(nics_service, nic)})
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler.py]: The update seemed more readable to me.

File: src/labeler/labeler.py [hunk 21]
```diff
@@ -43,88 +43,142 @@ def _get_host_interfaces_service(host_id):
     return _get_engine_service().hosts_service().host_service(host_id).nics_service()
 
 
-def _get_lldp_for_host(host_id):
+def _update_host_bonds(bonds, attachments, host_id):
+    host_service = _get_engine_service().hosts_service().host_service(host_id)
+    host_service.setup_networks(modified_bonds=bonds, modified_network_attachments=attachments)
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler.py]: I will do patch to reformat the code with yapf as Miguel suggested.

File: src/labeler/labeler.py [hunk 22]
```diff
@@ -43,88 +43,142 @@ def _get_host_interfaces_service(host_id):
...
+            attachments_to_update.extend(attachments)
+            next_bond_num += 1
+
+    if len(bonds_to_update) > 0:
+        _update_host_bonds(bonds_to_update, attachments_to_update, host_id)
+
+
 # TODO add diff check for the lldp from slaves
 def _get_lldp_for_bond_slaves(nics_service, nic):
     lldp_list = []
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler.py]: I am afraid that I would not be able to do it for every function.

File: src/labeler/labeler.py [hunk 23]
```diff
@@ -43,88 +43,142 @@ def _get_host_interfaces_service(host_id):
...
 
 
-def _get_lldp_for_host(host_id):
+def _update_host_bonds(bonds, attachments, host_id):
+    host_service = _get_engine_service().hosts_service().host_service(host_id)
+    host_service.setup_networks(modified_bonds=bonds, modified_network_attachments=attachments)
+    host_service.commit_net_config()
+
+
+def _get_lldp_for_host(host_id, without_bonds):
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler.py]: Done

File: src/labeler/labeler_utils.py [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+
+
+def update_bond_dict(bond_dict, tlvs, nic):
+    link_aggregation_properties = filter_link_aggregation(tlvs)
+    aggregation_port_id = (prop.value for prop in link_aggregation_properties if
+                           prop.name == PROPERTY_LINK_AGGREGATION_ID).next()
+    if aggregation_port_id is None:
+        return
+    slave_list = bond_dict.get(aggregation_port_id)
+    logging.info('Found active port aggregation group %s for nic %s', aggregation_port_id, nic.name)
```
[the remark this answers is not in the document]
Comment (almusil) [reply src/labeler/labeler_utils.py]: Done

File: src/labeler/labeler.py (hunk 15 above)
Comment (almusil) [reply src/labeler/labeler.py]: That might be something we should discuss. But I think it is safe as long we are not messing with management network.

File: Makefile (hunk 16 above)
Comment (almusil) [reply Makefile]: Done

**Pull request merged**